### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by msiebuhr** [msiebuhr/ucs#12]
While doing stress-testing with `go run ./cmd/ucs-bender -workers=5 -requests=100000`, I began getting logs indicating it's running against ulimits:

     server: namespace=default addr=127.0.0.1:56946 Error reading from cache: open /Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/cache5.0/default/ce/cefdfc072182654f163f5f0f9a621d72-9566c74d10037c4d7bbb0407d1e2c649.bin: too many open files
    server: namespace=default addr=127.0.0.1:56955 Error reading from cache: open /Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/cache5.0/default/ce/cefdfc072182654f163f5f0f9a621d72-9566c74d10037c4d7bbb0407d1e2c649.info: too many open files
    server: namespace=default addr=127.0.0.1:56952 Error reading from cache: open /Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/cache5.0/default/ce/cefdfc072182654f163f5f0f9a621d72-9566c74d10037c4d7bbb0407d1e2c649.info: too many open files
    server: namespace=default addr=127.0.0.1:56949 Error reading from cache: open /Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/cache5.0/default/ce/cefdfc072182654f163f5f0f9a621d72-9566c74d10037c4d7bbb0407d1e2c649.info: too many open files

And finally:

```
server: namespace=default Error accepting:  accept tcp [::]:8126: accept: too many open files
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13a9264]

goroutine 10 [running]:
github.com/msiebuhr/ucs.(*Server).Listener(0xc0000683c0, 0x151f800, 0xc00019c060, 0xc0001cc000, 0xf, 0xc0001a8f28)
	/Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/server.go:125 +0x1e4
github.com/msiebuhr/ucs.(*Server).Listen(0xc0000683c0, 0x151f780, 0xc000098008, 0x14aa351, 0x5, 0x0, 0x0)
	/Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/server.go:100 +0x1a3
created by main.main
	/Users/msiebuhr/Source/go/src/github.com/msiebuhr/ucs/cmd/ucs/main.go:115 +0x31a
exit status 2
```

**Comment by msiebuhr** [comment]
For good measure: It's in a branch where I'm experimenting with namespaces, so that may be related.

**Comment by msiebuhr** [comment]
Thinking it over, it should have some way of *ensuring* we'll do a GC on/after an upload. Current logic is to check `if (quota < size + newly-uploaded-item) { gc() }`, which means we'll do a GC on most uploads when we're hovering around the quota.

 * Run GC periodically (either timer or every X requests/connections)
 * Over-GC (say, 10% or just remove the oldest entry from *every* bucket)
 * Have a lock/queue, so we don't over-schedule GC's (we can upload small items a lot faster than we can GC them)
 * More fine-grained locking so GC doesn't block new uploads (goes with locked GC'ing, so we don't have something looking at files while another is deleting them...)

**Comment by msiebuhr** [comment]
After some thinking / digging it turned out to be caused by a missing `Close()` call when GC'ing.

Bumping the last comment to a new issue, as it really should be fixed too - see #13
